modules/standard/DistributedDeque.chpl

```diff
@@ -0,0 +1,1000 @@
+      return false;
+    }
+
+    // If we have a capacity, we must ensure that there is enough space for our
+    // operation ahead of time. We take a wait-free approach by performing a
+    // fetch-add on the queue size. If there is space, then we succeed within one
+    // communication round-trip. If we fail, then we make attempts to fix the queue
+    // size in a lock-free manner (for as long as it is over capacity). It should be
+    // noted that if we see that the queue size is negative, meaning a dequeue operation
+    // has over committed, we consequentially help them as well during the fetch-add loop.
+    if localThis.cap > 0 {
+      while true {
+        var size = queueSize.fetchAdd(1);
+
+        // Over capacity, fix before returning...
+        if size >= localThis.cap {
+          on this {
+            var readSize = queueSize.read();
+            // Attempt to fix, but yield to reduce potential contention and CPU hogging.
+            while readSize > this.cap && !queueSize.compareExchangeWeak(readSize, this.cap) {
+              chpl_task_yield();
+              readSize = queueSize.read();
+            }
+          }
+
+          // Fixed, return early...
+          if !DEQUE_NO_FREEZE then local { localThis.concurrentTasks.sub(1); }
+          return false;
+        }
+        // At this point, we either have a capacity and something between [0, cap],
+        // or we do not have a capacity and we have something greater than 0,
+        // either of which means we are safe.
+        else if size >= 0 {
+          break;
+        }
+        // If the size is negative, a dequeuer has overcommitted. Since there can
+        // be an element for us, we loop again.
+      }
+    }
+    // If we have a 'negative' capacity, we do not have one and can hold an
+    // infinite amount; this gives us a more optimized code path.
+    else {
+      while true {
+        var size = queueSize.fetchAdd(1);
+
+        // We have reserved a spot for our operation.
+        if size >= 0 {
+          break;
+        }
+        // If the size is negative, a dequeuer has overcommitted. Since there can
+        // be an element for us, we loop again.
+      }
+    }
+
+    return true;
+  }
+
+  /*
+    Syntactic sugar for `pushBack`.
+  */
+  proc add(elt : eltType) : bool {
+    return pushBack(elt);
+  }
+
+  /*
+    Syntactic sugar for `popFront`.
+  */
+  proc remove() : (bool, eltType) {
+    return popFront();
+  }
+
+  /*
+    Syntactic sugar for `pushBack`.
+  */
+  proc enqueue(elt : eltType) : bool {
+    return pushBack(elt);
+  }
+
+  /*
+    Syntactic sugar for `popFront`.
+  */
+  proc dequeue() : (bool, eltType) {
+    return popFront();
+  }
+
+  /*
+    Syntactic sugar for `pushBack`.
+  */
+  proc push(elt : eltType) : bool {
+    return pushBack(elt);
+  }
+
+  /*
+    Syntactic sugar for `popBack`.
+  */
+  proc pop() : (bool, eltType) {
+    return popBack();
+  }

```

Comment:
Since `Enqueue` pushes to the back, I'd expect that if I did `enqueue(); pop();` to get the same thing I just added, hence I have both use the same end.